lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -125,29 +153,70 @@ data SelectionCriteria = SelectionCriteria
 data SelectionResult = SelectionResult
     { inputsSelected
         :: !(NonEmpty (TxIn, TxOut))
+        -- ^ A (non-empty) list of selected inputs from the wallet's UTxO.
     , changeGenerated
         :: !(NonEmpty TokenBundle)
+        -- ^ A (non-empty) list of generated change outputs.
     , utxoRemaining
         :: !UTxOIndex
+        -- ^ UTxO remaining after performing a requested selection.
     }
     deriving (Eq, Show)
 
 -- | Represents the set of errors that may occur while performing a selection.
 --
-newtype SelectionError
+data SelectionError
     = BalanceInsufficient BalanceInsufficientError
+    | InsufficientMinCoinValues (NonEmpty InsufficientMinCoinValueError)
+    | UnableToConstructChange
+        -- ^ TODO: See if we can report how much ada is missing
+    deriving (Generic, Eq, Show)
 
 -- | Indicates that the balance of 'utxoAvailable' is insufficient to cover the
 --   balance of 'outputsToCover'.
 --
 data BalanceInsufficientError = BalanceInsufficientError
     { balanceAvailable
-        :: TokenBundle
+        :: !TokenBundle
       -- ^ The balance of 'utxoAvailable'.
     , balanceRequired
-        :: TokenBundle
+        :: !TokenBundle
       -- ^ The balance of 'outputsToCover'.
-    }
+    } deriving (Generic, Eq, Show)
+
+-- | Indicates that some of the specified outputs aren't valid and do not
+-- contain the minimum coin value expected by the protocol.
+--
+-- See also: 'prepareOutputs'.
+data InsufficientMinCoinValueError = InsufficientMinCoinValueError
+    { insufficientlyCoveredOutput
+        :: !TxOut
+        -- ^ The invalid output which doesn't hold enough coin.
+    , expectedMinCoinValue
+        :: !Coin
+        -- ^ The minimum coin value expected for this output.
+    } deriving (Generic, Eq, Show)
+
+-- | Prepare a set of outputs requested by users into valid Cardano outputs.
+-- That is, any output in Cardano needs to hold a minimum coin quantity (to
+-- prevent a certain kind of attack flooding the network with worthless UTxOs).
+--
+-- However, users do not typically specify a minimum ada value themselves.
+-- One would rather send '10 Apple' and not '10 Apple & 1.2 Ada'. Therefore,
+-- unless a coin value is explicitely specified, we do assign a coin value
+-- manually for each non-ada output. That value is the minimum value
+-- possible to make a particular output valid.
+prepareOutputsWith
+    :: (TokenMap -> Coin)
+    -> NonEmpty TxOut
+    -> NonEmpty TxOut
+prepareOutputsWith minCoinValueFor = fmap $ \out ->
+    out { Tx.tokens = augmentBundle (Tx.tokens out) }
+  where
+    augmentBundle bundle =
+        if TokenBundle.getCoin bundle == Coin 0
+        then bundle { coin = minCoinValueFor (view #tokens bundle) }
+        else bundle
```

Comment:
I think the code is much clearer now that the responsibility for preparing outputs has been moved out of `performSelection`. 👍🏻